### PR TITLE
Add permanent option context var 

### LINKED ISSUE #48: 💡 [REQUEST] - Context vars : Permanent option
### Start Date

_No response_

### Implementation PR

_No response_

### Reference Issues

_No response_

### Summary

This is an important feature to implement, we need to add a flag for context variables called "permanent" boolean. Currently a context var value is only kept during the current conversation and then gets lost. It would be good to add the "permanent" option allow the value to be stored in the subscriber object in a attribute called "context" (json).

### Basic Example

- Add "permanent" attribute to the context var entity (boolean)
- Add "context" attribute in the subscriber entity (JSON object)

### Drawbacks

We cannot typecheck the user context object.

### Unresolved questions

_No response_

### PR DESCRIPTION
**Pull request by medtaher123**
# Motivation

We need to add a flag for context variables called "permanent" boolean. Currently a context var value is only kept during the current conversation and then gets lost. It would be good to add the "permanent" option allow the value to be stored in the subscriber object in a attribute called "context" (json).

## Feature desciption:
There are two types of context variables in Hexabot: non-permanent and permanent.

### 1. Non-Permanent Context Variables

- **Temporary Storage:** Non-permanent context variables are stored in the conversation context.
- **Scope:** They are specific to the ongoing conversation and are lost once the conversation ends. This makes them ideal for storing temporary data that is relevant only during the current interaction.
- **Example:** If a bot asks a user for their preferred meeting time during a session, the bot may store the response as a non-permanent context variable to use in the rest of the conversation.

### 2. Permanent Context Variables

- **Persistent Storage:** Permanent context variables are stored in the subscriber context, which means they persist across multiple conversations.
- **Scope:** These variables retain their values even after a conversation has ended, allowing the bot to remember information about a user across different interactions.
- **Example:** If a user provides a preference for a certain feature, such as opting into notifications, the bot can store this preference permanently. The bot will retain this information for future use, without needing to ask the user again in subsequent conversations.

Fixes #48 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (Storybook)
- [x] I have added unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/src/chat/services/conversation.service.ts [hunk 1]
```diff
@@ -182,6 +206,13 @@ export class ConversationService extends BaseService<
           'Conversation Model : No conversation has been updated',
         );
       }
+
+      //TODO: add check if nothing changed don't update
+
+      this.subscriberService.updateOne(convo.sender, {
```
Comment (marrouchi) [inline api/src/chat/services/conversation.service.ts:212]:
```suggestion
      await this.subscriberService.updateOne(convo.sender, {
```

File: api/src/chat/services/conversation.service.ts [hunk 2]
```diff
@@ -79,13 +84,26 @@ export class ConversationService extends BaseService<
...
+      await this.contextVarService.find({
+        name: { $in: next.capture_vars.map((c) => c.context_var) },
+      })
+    ).reduce(
+      (acc, cv) => {
+        acc[cv.name] = cv;
+        return acc;
+      },
+      {} as Record<string, ContextVar>,
+    );
```
Comment (marrouchi) [inline api/src/chat/services/conversation.service.ts:105]: I suggest we add this method to the "contextVarService" : getContextVarsByBlock(next)

**Review by marrouchi — CHANGES_REQUESTED** [review]
Looks good @medtaher123 Thanks 👏  

I think you will need to adapt the unit tests.

File: api/src/chat/services/conversation.service.ts (hunk 1 above)
Comment (medtaher123) [reply api/src/chat/services/conversation.service.ts:212]: Yes, I'm aware of that. Thank you for pointing it out

Comment (medtaher123) [reply api/src/chat/services/conversation.service.ts:212]: Thank you for the suggestion. However, I think we don't need to await here, as making the function async allows the update to happen in the background. This way, we can continue processing other tasks without waiting for it to complete, which can improve performance, especially if the update isn't critical for the next steps

**Comment by medtaher123** [comment]
- [ ] Add "permanent" attribute to the context var entity 
- [ ] Add "context" attribute in the subscriber entity
- [ ] Capture permanent context vars in the subscriber
- [ ] correct existing unit tests
- [ ] add new corresponding unit tests
- [ ] Use the permanent context vars in the token replacement
- [ ] add unit tests
- [ ] add the permanent option in the frontentend contextVars page

File: api/src/chat/services/conversation.service.ts (hunk 1 above)
Comment (marrouchi) [reply api/src/chat/services/conversation.service.ts:202]: We need a await I think cause the next logic may need the subscriber to be up-to-date.

File: api/src/chat/dto/context-var.dto.ts [hunk 3]
```diff
@@ -20,6 +20,9 @@ export class ContextVarCreateDto {
   @IsNotEmpty()
   @IsString()
   name: string;
+
+  @ApiProperty({ description: 'Is context var permanent', type: Boolean })
```
Comment (marrouchi) [inline api/src/chat/dto/context-var.dto.ts:24]:
```suggestion
  @IsOptional()
  @IsBoolean()
  @ApiProperty({ description: 'Is context var permanent', type: Boolean })
```

**Review by marrouchi — CHANGES_REQUESTED** [review]

**Comment by marrouchi** [comment]
@medtaher123 Thank you 🙏  Can you please add a comment to the field permanent to explain what it does ?
Also, in the PR, add a detailed description of what a permanent context var is compared to a non permanent. Will be used in the documentation ;)

**Review by marrouchi — APPROVED** [review]

**Pull request merged**